### ISSUE TITLE
Distributed training jobs do not terminate properly if there is a crash

### ISSUE THREAD

**Issue by huihuifan** [pytorch/pytorch#17113]
## 🐛 Bug

If I use distributed training, sometimes one of the processes dies for a variety of reasons (maybe out of memory, a cuda runtime error, etc). In a single GPU job, the experiment would crash. In the distributed setting, the remaining processes often freeze and wait endlessly, so the job remains alive continues to consume GPU resources, even though it is actually dead. 

## Expected behavior

If one process in a distributed job crashes, all of the nodes in the distributed job should crash. 

## Environment

PyTorch 1.0

**Comment by huihuifan** [comment]
@mrshenli

**Comment by zhuyangda** [comment]
Me too, everything seems all right with the log has stoped.
It is difficult to determine what is the cause of the error.

**Comment by pietern** [comment]
@huihuifan Just to confirm, this happens when using the NCCL backend for torch.distributed?

**Comment by pietern** [comment]
If so, #17882 should solve it.

**Comment by Marcovaldong** [comment]
@pietern I got the same problem. Can we fix this error by changing the python code or reinstall the driver or other something?

**Comment by mrshenli** [comment]
@Marcovaldong Hopefully it is fixed in #22907, could you please give it a try?

**Comment by huihuifan** [comment]
@pietern @mrshenli , will try it out! thank you :)

**Comment by maye9999** [comment]
Same issue here. Using nccl backend, one job crashed without any error or logs and the whole program hanged. 
Pytorch 1.3.1

**Comment by vadimkantorov** [comment]
@mrshenli It would be good to have a controlled way to terminate all the processes, in case one of processes hits some critical error. A frequent example is OOM which is notoriously difficult to recover from in DDP setting

**Comment by bw4sz** [comment]
Is there update here? On a SLURM cluster ddp, one gpu dies and the job still is eating up the queue. Is there not a known solution

**Comment by vadimkantorov** [comment]
yep, some sort of timeout (maybe even with a default value say of 5 minutes!) and a forced kill of all spawned processes would be a great feature! @mrshenli

**Comment by stas00** [comment]
I second Vadim's feature request. Thank you!
